src/Voron/Data/BTrees/Tree.cs

```diff
@@ -1404,34 +1406,32 @@ public CompactTree CompactTreeFor(string key)
         
         public CompactTree CompactTreeFor(Slice key)
         {
-            _compactTrees ??= new Dictionary<Slice, CompactTree>(SliceComparer.Instance);
-
-            if (_compactTrees.TryGetValue(key, out var compactTree) == false)
+            if (_compactTrees == null)
+            {
+                // Since the Weak Small Set will lose items if more than 128 items gets added,
+                // we will still be adding the compact trees to a dictionary; however, since
+                // the dictionary will get searched many times over the lifetime of the transaction
+                // the overhead caused by storing in two different data structures gets amortized.
+                _compactTrees = new Dictionary<Slice, CompactTree>(SliceComparer.Instance);
+                _compactTreesLocator = new WeakSliceSmallSet<CompactTree>(128);
```

Comment:
We always need this together, so lets drop the weak reference from here and put the dictionary inside

Reply:
Not necessarily. For instance, we have a pretty fast page locator already, but this could act as a page locator to.

Reply:
Current implementation does not insert into the backing storage unless it requires to do so. Much more efficient than what we have now.